mozilla_bitbar_devicepool/test_run_manager.py

```diff
@@ -88,135 +68,137 @@ def get_bitbar_test_stats(self, project_name, project_config):
-        stats = {
-            'COUNT': device_group_count,
-            'IDLE': 0,
-            'OFFLINE_DEVICES': offline_devices,
-            'OFFLINE': len(offline_devices),
-            'DISABLED': device_group_count - len(enabled_devices),
-            'FINISHED': 0,
-            'RUNNING': 0,
-            'WAITING': 0,
-            }
-
-        for i, test_run in enumerate(bitbar_test_runs[project_name]):
-            test_run_id = test_run['id']
+
+        stats = CACHE['projects'][project_name]['stats']
+        stats['OFFLINE_DEVICES'] = offline_devices
+        stats['OFFLINE'] = len(offline_devices)
+        stats['DISABLED'] = device_group_count - len(enabled_devices)
+        stats['RUNNING'] = 0
+        stats['WAITING'] = 0
+
+        for test_run in bitbar_test_runs[project_name]:
             test_run_state = test_run['state']
-            # Refresh test_run from bitbar and update the cache.
-            test_run = bitbar_test_runs[project_name][i] = get_test_run(project_id, test_run_id)
             stats[test_run_state] += 1
-            if test_run_state == 'FINISHED':
-                logger.info('{:10s} test run {} finished'.format(device_group_name, test_run['id']))
-                del bitbar_test_runs[project_name][i]
-                if self.delete_bitbar_tests:
-                    delete_test_run(project_id, test_run_id)
 
         stats['IDLE'] = (stats['COUNT'] -
                          stats['DISABLED'] -
                          stats['OFFLINE'] -
                          stats['RUNNING'])
 
-        return stats
+    def handle_queue(self, project_name, projects_config):
+        logger.info("thread starting")
+        stats = CACHE['projects'][project_name]['stats']
+
+        while self.state == 'RUNNING':
+            project_config = projects_config[project_name]
+            device_group_name = project_config['device_group_name']
+            additional_parameters = project_config['additional_parameters']
+            worker_type = additional_parameters.get('TC_WORKER_TYPE')
+
+            if stats['OFFLINE'] or stats['DISABLED']:
+                logger.warning('{:10s} DISABLED {} OFFLINE {} {}'.format(
+                    device_group_name,
+                    stats['DISABLED'],
+                    stats['OFFLINE'],
+                    ', '.join(stats['OFFLINE_DEVICES'])))
+
+            bitbar_device_group = CACHE['device_groups'][device_group_name]
+            bitbar_device_group_count = bitbar_device_group['deviceCount']
+            taskcluster_provisioner_id = projects_config['defaults']['taskcluster_provisioner_id']
+
+            # create enough tests to service either the pending tasks or twice the number
+            # of the devices in the group (whichever is smaller).
+            pending_tasks = get_taskcluster_pending_tasks(taskcluster_provisioner_id, worker_type)
+            jobs_to_start = min(pending_tasks, 2*bitbar_device_group_count) - stats['WAITING']
```

Comment:
Let's change this from 2*bitbar_device_group_count to bitbar_device_group_count and just create enough to keep the device group busy but not too busy.

Reply:
Done.